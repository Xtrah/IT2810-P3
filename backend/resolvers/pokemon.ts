/* eslint-disable no-useless-catch */

import { Pokemon, PokemonAttributes } from '../models/pokemon';

interface QueryOptions {
  name: { $regex: RegExp };
  types?: string;
}

const pokemonResolver = {
  /*
    Return all pokemons or query name or type.
    Sort alphabetically on name by default. Can sort descending.
  */
  pokemons: async (args: {
    name: string;
    sortDescending: boolean;
    type: string;
  }) => {
    try {
      /*
        Defining name query is optional.
        Query doesn't have to be exact as name.
        Query is case insensitive. 
      */
      const queryOptions: QueryOptions = {
        name: { $regex: new RegExp(args.name || '', 'i') },
      };
      // Query on type is optional
      if (args.type) {
        queryOptions.types = args.type.toLowerCase();
      }

      // Default sort is alphabetically. Can query descending.
      const sortOptions = { name: args.sortDescending ? -1 : 1 };
      const pokemons = await Pokemon.find(queryOptions)
        .sort(sortOptions)
        .exec();
      return pokemons;
    } catch (err) {
      throw err;
    }
  },
  // Create a pokemon with certain attributes
  createPokemon: async (args: { pokemonInput: PokemonAttributes }) => {
    const { name, description, types, height, weight, imageUrl } =
      args.pokemonInput;

    const pokemon = Pokemon.build({
      name,
      description,
      types,
      height,
      weight,
      imageUrl,
    });

    try {
      const result = await pokemon.save();
      return result;
    } catch (err) {
      throw err;
    }
  },
};

export default pokemonResolver;
