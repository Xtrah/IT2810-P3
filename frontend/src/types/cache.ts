import { InMemoryCache, makeVar } from '@apollo/client';

export const cache: InMemoryCache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        pokemonFilter: {
          read() {
            return pokemonFilterVar();
          },
        },
      },
    },
  },
});

export interface PokemonFilter {
  type: string;
  sortDescending: boolean;
}

/**
 * PokemonFilterVar handles the filter variables.
 *
 * It allows for saving and retireval of filter states in the cache.
 */
export const pokemonFilterVar = makeVar<PokemonFilter>({
  type: '',
  sortDescending: false,
});

// Here we set the initial values for the filter.
