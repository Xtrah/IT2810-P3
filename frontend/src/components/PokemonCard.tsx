import { Box, Center, Text, Image, Link } from '@chakra-ui/react';
import { Link as RouterLink } from 'react-router-dom';
import { PokemonLimited } from '../types/pokemon';
import getGradientByType from '../utils/getGradientByType';
import getIconByType from '../utils/getIconByType';

interface Props {
  pokemon: PokemonLimited;
}

// PokemonCard is used as list item in search results
const PokemonCard = ({ pokemon }: Props) => (
  <Link as={RouterLink} to={`/pokemon/${pokemon._id}`}>
    <Center
      key={pokemon._id}
      borderRadius="xl"
      mt={2}
      mb={2}
      h="100px"
      bgGradient={getGradientByType(pokemon.types[0])}
      _hover={{
        opacity: '0.7',
        transition: '0.5s',
      }}
    >
      <Box mr={5}>
        <Text color="white">
          {pokemon.name}
          {pokemon.types.map((type) => <Image
              key={type}
              ml={2}
              display="inline"
              borderRadius="xl"
              boxSize="20px"
              objectFit="cover"
              src={getIconByType(type)}
              alt={pokemon.name}/>)
            }
        </Text>
      </Box>
      <Image
        borderRadius="xl"
        boxSize="50px"
        objectFit="cover"
        src={pokemon.imageUrl}
        // Fallback to Bulbasaur if image doesn't load
        fallbackSrc="https://assets.pokemon.com/assets/cms2/img/pokedex/full/001.png"
        alt={pokemon.name}
      />
    </Center>
  </Link>
);

export default PokemonCard;
