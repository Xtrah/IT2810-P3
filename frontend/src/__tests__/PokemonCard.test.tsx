import { ChakraProvider } from '@chakra-ui/react';
import renderer from 'react-test-renderer';
import PokemonCard from '../components/PokemonCard';
import { PokemonLimited } from '../types/pokemon';

// Inspired by https://blog.openreplay.com/react-snapshot-testing-with-jest-an-introduction-with-examples

describe('<PokemonCard />', () => {
  it('should match snapshot with data', () => {
    const pokemonData: PokemonLimited = {
      _id: 'randomid',
      name: 'Charizard',
      description: 'Fireeee',
      types: ['fire'],
      imageUrl: 'https://somerandomurl123123123.no',
    };

    const elem = renderer
      .create(
        <ChakraProvider>
          <PokemonCard pokemon={pokemonData} />
        </ChakraProvider>
      )
      .toJSON();

    expect(elem).toMatchSnapshot();
  });
});
