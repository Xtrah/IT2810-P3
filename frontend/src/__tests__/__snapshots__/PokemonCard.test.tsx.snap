// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonCard /> should match snapshot with data 1`] = `
Array [
  <a
    className="chakra-link css-f4h6uy"
    href="/pokemon/randomid"
    onClick={[Function]}
  >
    <div
      className="css-166di66"
    >
      <div
        className="css-1g3r3q7"
      >
        <p
          className="chakra-text css-79wky"
        >
          Charizard
          <img
            alt="Charizard"
            className="chakra-image css-h13pur"
            src="fire.svg"
          />
        </p>
      </div>
      <img
        alt="Charizard"
        className="chakra-image__placeholder css-8dgyiu"
        src="https://assets.pokemon.com/assets/cms2/img/pokedex/full/001.png"
      />
    </div>
  </a>,
  <span />,
]
`;
